### PR TITLE
[Feature] support url_encode and url_decode functions

### LINKED ISSUE #26724: [feature] url_encode & url_decode
It's good to have url_encode and url_decode function that translates a string into [application/x-www-form-urlencoded](https://www.w3.org/TR/html4/interact/forms.html#h-17.13.4.1) format. 

Examples:
https://docs.databricks.com/sql/language-manual/functions/url_encode.html
https://docs.databricks.com/sql/language-manual/functions/url_decode.html
https://trino.io/docs/current/functions/url.html?highlight=url_encode#url_encode
https://trino.io/docs/current/functions/url.html?highlight=url_encode#url_decode

## syntax

```
url_encode(str)
```

## input
string expression

## output
string

## examples
```
SELECT url_encode('https://docs.starrocks.io/en-us/latest/quick_start/Deploy');
https%3A%2F%2Fdocs.starrocks.io%2Fen-us%2Flatest%2Fquick_start%2FDeploy
```

# url_decode

## syntax

```
url_decode(str)
```

## input
string expression

## output
string

## examples
```
SELECT url_decode(https%3A%2F%2Fdocs.starrocks.io%2Fen-us%2Flatest%2Fquick_start%2FDeploy');
https://docs.starrocks.io/en-us/latest/quick_start/Deploy 
```

### PR DESCRIPTION
**Pull request by xuzifu666**
Fixes #26724

## What type of PR is this:
- [ ] BugFix
- [x] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [x] Doc
- [ ] Tool

## Checklist:
- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [x] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [ ] I have checked the version labels which the pr will be auto-backported to the target branch
  - [ ] 3.1
  - [ ] 3.0
  - [ ] 2.5
  - [ ] 2.4

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

File: docs/TOC.md [hunk 1]
```diff
@@ -620,6 +620,8 @@
       + [money_format](./sql-reference/sql-functions/string-functions/money_format.md)
       + [null_or_empty](./sql-reference/sql-functions/string-functions/null_or_empty.md)
       + [parse_url](./sql-reference/sql-functions/string-functions/parse_url.md)
+      + [url_encode](./sql-reference/sql-functions/string-functions/url_encode.md)
```
Comment (wangsimo0) [inline docs/TOC.md:623]: need to be added after upper

Comment (xuzifu666) [reply docs/TOC.md:623]: get it

**Review by murphyatwork — DISMISSED** [review]

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

File: test/sql/test_time_fn/T/test_time_fn [hunk 2]
```diff
@@ -222,6 +222,14 @@ values ('2010-11-30 23:59:59', '2010-11-30 20:58:59', 'hour'),
...
 select date_diff(d0,d1,t) from test_data_diff order by t;
 
+-- name: test_url_encode
+select url_encode('https://docs.starrocks.io/en-us/latest/quick_start/Deploy');
+select url_encode(NULL);
+
+--name: test_url_decode
+select url_decode('https%3A%2F%2Fdocs.starrocks.io%2Fen-us%2Flatest%2Fquick_start%2FDeploy');
+select url_decode(NULL);
+
```
Comment (ucasfl) [inline test/sql/test_time_fn/T/test_time_fn:232]: How about URL includes Chinese?

File: test/sql/test_time_fn/R/test_time_fn [hunk 3]
```diff
@@ -787,6 +787,26 @@ select date_diff(d0,d1,t) from test_data_diff order by t;
...
+--name: test_url_decode
+select url_decode('https%3A%2F%2Fdocs.starrocks.io%2Fen-us%2Flatest%2Fquick_start%2FDeploy');
+-- result:
+https://docs.starrocks.io/en-us/latest/quick_start/Deploy
+-- !result
+select url_decode(NULL);
+-- result:
+None
+-- !result
+
```
Comment (silverbullet233) [inline test/sql/test_time_fn/R/test_time_fn:809]: these cases should not be added into test_time_fn file, better to create a new file under test/sql/test_function

Comment (xuzifu666) [reply test/sql/test_time_fn/R/test_time_fn:809]: ok，i move it to test_string function

File: test/sql/test_time_fn/T/test_time_fn (hunk 2 above)
Comment (xuzifu666) [reply test/sql/test_time_fn/T/test_time_fn]: It can not in Chinese，because url decode would resolve as other chars

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

File: test/sql/test_time_fn/R/test_time_fn (hunk 3 above)
Comment (xuzifu666) [reply test/sql/test_time_fn/R/test_time_fn]: ok，i move it to string function sql testcase @silverbullet233

File: test/sql/test_time_fn/T/test_time_fn (hunk 2 above)
Comment (xuzifu666) [reply test/sql/test_time_fn/T/test_time_fn]: ex：

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

File: be/src/exprs/string_functions.cpp [hunk 4]
```diff
@@ -2075,6 +2079,58 @@ StatusOr<ColumnPtr> StringFunctions::unhex(FunctionContext* context, const starr
...
+    int i, ii;
+    for (i = 0; i < value.length(); i++) {
+        if (int(value[i]) == 37) {
+            sscanf(value.substr(i + 1, 2).c_str(), "%x", &ii);
+            ch = static_cast<char>(ii);
+            ret += ch;
+            i = i + 2;
+        } else {
+            ret += value[i];
+        }
```
Comment (ucasfl) [inline be/src/exprs/string_functions.cpp:2125]:
```suggestion
    short x;
    size_t length = value.length();
    for (size_t i = 0; i < length; i++) {
        if (int(value[i]) == 37) {
            sscanf(value.substr(i + 1, 2).c_str(), "%x", &x);
            ch = static_cast<char>(x);
            ret += ch;
            i += 2;
        } else {
            ret += value[i];
        }
```

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

File: docs/sql-reference/sql-functions/string-functions/url_decode.md [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+- `str`: the string to decode. If `str` is not a string type, it will try implicit cast first.
+
+## Return values
+
+Return an encode string.
+
+## Examples
+
+```plaintext
+mysql> select url_decode('https%3A%2F%2Fdocs.starrocks.io%2Fen-us%2Flatest%2Fquick_start%2FDeploy');
```
Comment (fzhedu) [inline docs/sql-reference/sql-functions/string-functions/url_decode.md:24]: what if url contains non-ENG ? for example Chinese?

Comment (xuzifu666) [reply docs/sql-reference/sql-functions/string-functions/url_decode.md:24]: it's ok
for example：
mysql> select url_encode('[https://docs.starrocks数据库.io/en-us/latest/quick_start/Deploy?name=快速](https://docs.xn--starrocks-ev0ri14bh0f.io/en-us/latest/quick_start/Deploy?name=%E5%BF%AB%E9%80%9F)');
+--------------------------------------------------------------------------------------------------------------------------------+
| url_encode('[https://docs.starrocks数据库.io/en-us/latest/quick_start/Deploy?name=快速](https://docs.xn--starrocks-ev0ri14bh0f.io/en-us/latest/quick_start/Deploy?name=%E5%BF%AB%E9%80%9F)') |
+--------------------------------------------------------------------------------------------------------------------------------+
| https%3A%2F%2Fdocs.starrocks%E6%95%B0%E6%8D%AE%E5%BA%93.io%2Fen-us%2Flatest%2Fquick_start%2FDeploy%3Fname%3D%E5%BF%AB%E9%80%9F |
+--------------------------------------------------------------------------------------------------------------------------------+
1 row in set (0.01 sec)

mysql> select url_decode('https%3A%2F%2Fdocs.starrocks%E6%95%B0%E6%8D%AE%E5%BA%93.io%2Fen-us%2Flatest%2Fquick_start%2FDeploy%3Fname%3D%E5%BF%AB%E9%80%9F');
+----------------------------------------------------------------------------------------------------------------------------------------------+
| url_decode('https%3A%2F%2Fdocs.starrocks%E6%95%B0%E6%8D%AE%E5%BA%93.io%2Fen-us%2Flatest%2Fquick_start%2FDeploy%3Fname%3D%E5%BF%AB%E9%80%9F') |
+----------------------------------------------------------------------------------------------------------------------------------------------+
| [https://docs.starrocks数据库.io/en-us/latest/quick_start/Deploy?name=快速](https://docs.xn--starrocks-ev0ri14bh0f.io/en-us/latest/quick_start/Deploy?name=%E5%BF%AB%E9%80%9F) |
+----------------------------------------------------------------------------------------------------------------------------------------------+

@fzhedu

File: be/src/exprs/string_functions.cpp [hunk 6]
```diff
@@ -2075,6 +2079,58 @@ StatusOr<ColumnPtr> StringFunctions::unhex(FunctionContext* context, const starr
...
+DEFINE_STRING_UNARY_FN_WITH_IMPL(url_decodeImpl, str) {
+    return StringFunctions::url_decode_func(str.to_string());
+}
+
+std::string StringFunctions::url_decode_func(const std::string& value) {
+    std::string ret;
+    char ch;
+    int i, ii;
+    for (i = 0; i < value.length(); i++) {
+        if (int(value[i]) == 37) {
```
Comment (LiShuMing) [inline be/src/exprs/string_functions.cpp:2118]: `value[i] == '%'`

File: be/src/exprs/string_functions.cpp [hunk 7]
```diff
@@ -2075,6 +2079,58 @@ StatusOr<ColumnPtr> StringFunctions::unhex(FunctionContext* context, const starr
...
+            sscanf(value.substr(i + 1, 2).c_str(), "%x", &ii);
```
Comment (LiShuMing) [inline be/src/exprs/string_functions.cpp:2119]: what about use `      std::istringstream is(in.substr(i + 1, 2))` ?

Comment (LiShuMing) [reply be/src/exprs/string_functions.cpp:2119]: What if `value.substr(i + 1, 2).c_str()` is illegal hex? return null or throw exception?

**Review by fzhedu — APPROVED** [review]
LGTM

File: be/src/exprs/string_functions.cpp (hunk 6 above)
Comment (xuzifu666) [reply be/src/exprs/string_functions.cpp]: done

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

File: be/src/exprs/string_functions.cpp [hunk 8]
```diff
@@ -2075,6 +2079,58 @@ StatusOr<ColumnPtr> StringFunctions::unhex(FunctionContext* context, const starr
...
+    return StringFunctions::url_encode_func(str.to_string());
+}
+
+std::string StringFunctions::url_encode_func(const std::string& value) {
+    std::ostringstream escaped;
+    escaped.fill('0');
+    escaped << std::hex;
+
+    for (std::string::const_iterator i = value.begin(), n = value.end(); i != n; ++i) {
+        std::string::value_type c = (*i);
```
[the remark this answers is not in the document]
Comment (ucasfl) [reply be/src/exprs/string_functions.cpp:2092]: The code does not resolved. @xuzifu666

File: be/src/exprs/string_functions.cpp (hunk 4 above)
Comment (ucasfl) [reply be/src/exprs/string_functions.cpp]: Not resolved. @xuzifu666

File: be/src/exprs/string_functions.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (xuzifu666) [reply be/src/exprs/string_functions.cpp]: done

File: be/src/exprs/string_functions.cpp (hunk 4 above)
Comment (xuzifu666) [reply be/src/exprs/string_functions.cpp]: done

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

**Review by fzhedu — APPROVED** [review]
LGTM

**Review by ucasfl — APPROVED** [review]
LGTM

**Review by silverbullet233 — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

**Pull request merged**